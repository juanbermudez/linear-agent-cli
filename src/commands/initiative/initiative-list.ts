import { Command } from "@cliffy/command"
import { Table } from "@cliffy/table"
import { listInitiatives, listUsers } from "../../utils/linear.ts"
import { formatRelativeTime, truncate } from "../../utils/display.ts"
import { muted } from "../../utils/styling.ts"
import { error as errorColor } from "../../utils/styling.ts"

interface ListOptions {
  status?: string
  owner?: string
  limit?: number
  json?: boolean
  format?: string
}

export const listCommand = new Command()
  .name("list")
  .description("List initiatives")
  .option("-s, --status <status:string>", "Filter by status ID or name")
  .option("-o, --owner <owner:string>", "Filter by owner ID or display name")
  .option("-l, --limit <limit:number>", "Max number of initiatives (default: 50)")
  .option("-j, --json", "Output result as JSON")
  .option("--format <format:string>", "Output format: text|json")
  .action(async (options: ListOptions) => {
    const useJson = options.json || options.format === "json"

    let statusId: string | undefined
    let ownerId: string | undefined

    // Resolve status if provided (not implemented in this version - would need listInitiativeStatuses)
    if (options.status) {
      statusId = options.status
    }

    // Resolve owner if provided
    if (options.owner) {
      if (
        /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(
          options.owner,
        )
      ) {
        ownerId = options.owner
      } else {
        const users = await listUsers()
        const user = users.find((u) =>
          u.displayName.toLowerCase() === options.owner!.toLowerCase()
        )
        if (user) {
          ownerId = user.id
        } else {
          const errorMsg = `User '${options.owner}' not found`
          if (useJson) {
            console.error(
              JSON.stringify({
                success: false,
                error: { code: "NOT_FOUND", message: errorMsg },
              }, null, 2),
            )
          } else {
            console.error(errorColor(`Error: ${errorMsg}`))
          }
          Deno.exit(1)
        }
      }
    }

    const { Spinner } = await import("@std/cli/unstable-spinner")
    const showSpinner = !useJson && Deno.stdout.isTerminal()
    const spinner = showSpinner
      ? new Spinner({ message: "Fetching initiatives..." })
      : null
    spinner?.start()

    try {
      const initiatives = await listInitiatives({
        statusId,
        ownerId,
        limit: options.limit,
      })

      spinner?.stop()

      if (useJson) {
        console.log(
          JSON.stringify({
            initiatives: initiatives.map((i) => ({
              id: i.id,
              name: i.name,
              slugId: i.slugId,
              url: i.url,
              status: i.status
                ? { id: i.status.id, name: i.status.name, type: i.status.type }
                : null,
              owner: i.owner
                ? { id: i.owner.id, name: i.owner.displayName }
                : null,
              targetDate: i.targetDate,
              createdAt: i.createdAt,
              updatedAt: i.updatedAt,
            })),
            count: initiatives.length,
          }, null, 2),
        )
        return
      }

      if (initiatives.length === 0) {
        console.log(muted("No initiatives found"))
        return
      }

      const rows = initiatives.map((i) => {
        const status = i.status ? i.status.name : "-"
        const owner = i.owner ? i.owner.displayName : "-"
        const targetDate = i.targetDate
          ? formatRelativeTime(i.targetDate)
          : "-"

        return [
          i.slugId,
          i.name,
          status,
          owner,
          targetDate,
        ]
      })

      const terminalWidth = Deno.consoleSize().columns
      const fixedWidth = 10 + 20 + 15 + 20 // ID + Status + Owner + Target
      const nameWidth = Math.max(20, terminalWidth - fixedWidth - 10)

      const table = new Table()
        .header(["ID", "Name", "Status", "Owner", "Target"])
        .body(
          rows.map(([id, name, status, owner, target]) => {
            const truncatedName = truncate(name, nameWidth)
            return [id, truncatedName, status, owner, target]
          }),
        )
        .border()

      table.render()
    } catch (err) {
      spinner?.stop()
      if (useJson) {
        console.error(
          JSON.stringify({
            success: false,
            error: { code: "API_ERROR", message: err.message },
          }, null, 2),
        )
      } else {
        console.error(errorColor(`Error: ${err.message}`))
      }
      Deno.exit(1)
    }
  })
