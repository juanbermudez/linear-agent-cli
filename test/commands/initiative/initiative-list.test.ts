import { snapshotTest } from "@cliffy/testing"
import { listCommand } from "../../../src/commands/initiative/initiative-list.ts"
import {
  commonDenoArgs,
  setupMockLinearServer,
} from "../../utils/test-helpers.ts"

// Test help output
await snapshotTest({
  name: "Initiative List Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs: commonDenoArgs,
  async fn() {
    await listCommand.parse()
  },
})

// Test listing all initiatives
await snapshotTest({
  name: "Initiative List Command - All Initiatives",
  meta: import.meta,
  colors: false,
  args: ["--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetInitiatives",
        response: {
          data: {
            initiatives: {
              nodes: [
                {
                  id: "init-1",
                  name: "Q1 2025 Platform Goals",
                  slugId: "q1-2025-platform-goals",
                  url:
                    "https://linear.app/test/initiative/q1-2025-platform-goals",
                  description: "Focus on platform stability and performance",
                  status: {
                    id: "status-active",
                    name: "Active",
                    type: "started",
                  },
                  owner: {
                    name: "alice",
                    displayName: "Alice Smith",
                  },
                  targetDate: "2025-03-31",
                  createdAt: "2024-01-01T10:00:00Z",
                  updatedAt: "2024-01-15T14:30:00Z",
                },
                {
                  id: "init-2",
                  name: "API Modernization",
                  slugId: "api-modernization",
                  url: "https://linear.app/test/initiative/api-modernization",
                  description: "Modernize our API infrastructure",
                  status: {
                    id: "status-planned",
                    name: "Planned",
                    type: "planned",
                  },
                  owner: {
                    name: "bob",
                    displayName: "Bob Jones",
                  },
                  targetDate: "2025-06-30",
                  createdAt: "2024-01-05T09:00:00Z",
                  updatedAt: "2024-01-10T11:20:00Z",
                },
                {
                  id: "init-3",
                  name: "Customer Experience Improvements",
                  slugId: "customer-experience-improvements",
                  url:
                    "https://linear.app/test/initiative/customer-experience-improvements",
                  description: null,
                  status: null,
                  owner: null,
                  targetDate: null,
                  createdAt: "2024-01-10T15:00:00Z",
                  updatedAt: "2024-01-10T15:00:00Z",
                },
              ],
            },
          },
        },
      },
    ])

    try {
      await listCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test listing with JSON output
await snapshotTest({
  name: "Initiative List Command - JSON Output",
  meta: import.meta,
  colors: false,
  args: ["--json", "--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetInitiatives",
        response: {
          data: {
            initiatives: {
              nodes: [
                {
                  id: "init-1",
                  name: "Q1 2025 Platform Goals",
                  slugId: "q1-2025-platform-goals",
                  url:
                    "https://linear.app/test/initiative/q1-2025-platform-goals",
                  description: "Focus on platform stability",
                  status: {
                    id: "status-active",
                    name: "Active",
                    type: "started",
                  },
                  owner: {
                    name: "alice",
                    displayName: "Alice Smith",
                  },
                  targetDate: "2025-03-31",
                  createdAt: "2024-01-01T10:00:00Z",
                  updatedAt: "2024-01-15T14:30:00Z",
                },
              ],
            },
          },
        },
      },
    ])

    try {
      await listCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test filtering by status
await snapshotTest({
  name: "Initiative List Command - Filter By Status",
  meta: import.meta,
  colors: false,
  args: ["--status", "started", "--json", "--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetInitiatives",
        variables: { filter: { status: { type: { eq: "started" } } } },
        response: {
          data: {
            initiatives: {
              nodes: [
                {
                  id: "init-1",
                  name: "Q1 2025 Platform Goals",
                  slugId: "q1-2025-platform-goals",
                  url:
                    "https://linear.app/test/initiative/q1-2025-platform-goals",
                  description: "Focus on platform stability",
                  status: {
                    id: "status-active",
                    name: "Active",
                    type: "started",
                  },
                  owner: {
                    name: "alice",
                    displayName: "Alice Smith",
                  },
                  targetDate: "2025-03-31",
                  createdAt: "2024-01-01T10:00:00Z",
                  updatedAt: "2024-01-15T14:30:00Z",
                },
              ],
            },
          },
        },
      },
    ])

    try {
      await listCommand.parse()
    } finally {
      await cleanup()
    }
  },
})

// Test empty initiative list
await snapshotTest({
  name: "Initiative List Command - No Initiatives Found",
  meta: import.meta,
  colors: false,
  args: ["--no-color"],
  denoArgs: commonDenoArgs,
  async fn() {
    const { cleanup } = await setupMockLinearServer([
      {
        queryName: "GetInitiatives",
        response: {
          data: {
            initiatives: {
              nodes: [],
            },
          },
        },
      },
    ])

    try {
      await listCommand.parse()
    } finally {
      await cleanup()
    }
  },
})
