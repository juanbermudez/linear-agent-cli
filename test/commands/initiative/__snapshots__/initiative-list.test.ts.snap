export const snapshot = {};

snapshot[`Initiative List Command - Help Text 1`] = `
stdout:
"
Usage: list

Description:

  List initiatives

Options:

  -h, --help              - Show this help.                          
  -s, --status  <status>  - Filter by status ID or name              
  -o, --owner   <owner>   - Filter by owner ID or display name       
  -l, --limit   <limit>   - Max number of initiatives (default: 50)  
  --plain                 - Disable colored output                   
  --no-color              - Disable colored output                   
  -j, --json              - Output result as JSON                    
  --format      <format>  - Output format: text|json                 

"
stderr:
""
`;

snapshot[`Initiative List Command - All Initiatives 1`] = `
stdout:
'{
  "initiatives": [
    {
      "id": "init-1",
      "name": "Q1 2025 Platform Goals",
      "slugId": "q1-2025-platform-goals",
      "url": "https://linear.app/test/initiative/q1-2025-platform-goals",
      "status": "started",
      "owner": {
        "id": "user-alice",
        "name": "Alice Smith"
      },
      "targetDate": "2025-03-31",
      "createdAt": "2024-01-01T10:00:00Z",
      "updatedAt": "2024-01-15T14:30:00Z"
    },
    {
      "id": "init-2",
      "name": "API Modernization",
      "slugId": "api-modernization",
      "url": "https://linear.app/test/initiative/api-modernization",
      "status": "planned",
      "owner": {
        "id": "user-bob",
        "name": "Bob Jones"
      },
      "targetDate": "2025-06-30",
      "createdAt": "2024-01-05T09:00:00Z",
      "updatedAt": "2024-01-10T11:20:00Z"
    },
    {
      "id": "init-3",
      "name": "Customer Experience Improvements",
      "slugId": "customer-experience-improvements",
      "url": "https://linear.app/test/initiative/customer-experience-improvements",
      "status": null,
      "owner": null,
      "targetDate": null,
      "createdAt": "2024-01-10T15:00:00Z",
      "updatedAt": "2024-01-10T15:00:00Z"
    }
  ],
  "count": 3
}
'
stderr:
""
`;

snapshot[`Initiative List Command - JSON Output 1`] = `
stdout:
'{
  "initiatives": [
    {
      "id": "init-1",
      "name": "Q1 2025 Platform Goals",
      "slugId": "q1-2025-platform-goals",
      "url": "https://linear.app/test/initiative/q1-2025-platform-goals",
      "status": "started",
      "owner": {
        "id": "user-alice",
        "name": "Alice Smith"
      },
      "targetDate": "2025-03-31",
      "createdAt": "2024-01-01T10:00:00Z",
      "updatedAt": "2024-01-15T14:30:00Z"
    }
  ],
  "count": 1
}
'
stderr:
""
`;

snapshot[`Initiative List Command - Filter By Status 1`] = `
stdout:
'{
  "initiatives": [
    {
      "id": "init-1",
      "name": "Q1 2025 Platform Goals",
      "slugId": "q1-2025-platform-goals",
      "url": "https://linear.app/test/initiative/q1-2025-platform-goals",
      "status": "started",
      "owner": {
        "id": "user-alice",
        "name": "Alice Smith"
      },
      "targetDate": "2025-03-31",
      "createdAt": "2024-01-01T10:00:00Z",
      "updatedAt": "2024-01-15T14:30:00Z"
    }
  ],
  "count": 1
}
'
stderr:
""
`;

snapshot[`Initiative List Command - No Initiatives Found 1`] = `
stdout:
"No initiatives found
"
stderr:
""
`;
